common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ErrorHandler.java

```diff
@@ -105,16 +105,16 @@ public boolean shouldRetryError(Throwable t) {
         return false;
       }
 
-      String errorStackTrace = Throwables.getStackTraceAsString(t);
       // If the block is too late or stale block push, there is no need to retry it
-      return !errorStackTrace.contains(TOO_LATE_OR_STALE_BLOCK_PUSH_MESSAGE_SUFFIX);
+      String msg = t.getMessage();
+      return !(msg != null && msg.contains(TOO_LATE_OR_STALE_BLOCK_PUSH_MESSAGE_SUFFIX));
```

Comment:
I really think the way to matching the target string is kind of fragile. It potentially assumes the structure of the error. But I think we can not make sure that the error structure would never change.

So I'm thinking about using an enum type to represent different push results and reply with a normal RPC message.

WDYT?

Reply:
I actually thought of the same during the performance evaluation.
We are currently using the exception as some sort of RPC message to deliver these 2 potentially frequent messages, i.e. too late and block collision.
It incurs overhead on both server side and client side for initializing the exception and examine the content of the exception, as well as this issue you pointed out.
It is ideally much better to make these 2 a proper RPC response, e.g. responding an actual payload with a byte instead of an empty ByteBuffer.
We can still keep the response in the case of a success push and merge of a block as an empty ByteBuffer to reduce the cost of decoding ByteBuffer for every success block push.
This however is a protocol change, and I'm not sure we will be able to change it before the 3.2 RC date.

cc @gengliangwang

Reply:
Well, it is not really a protocol change.
We are just switching from RPCFailure to RPCResponse using a non-empty ByteBuffer.

Reply:
@Victsm I planned to cut RC on Aug 10th.

Reply:
Had some discussions related to making this a proper RPC response with @Victsm in the past and I think as well that it is a better approach.  One thing to note though is that if we do this, we will have to make changes on the client to invoke `listener.onFailure()` callback when the RPCResponse represents these exception types because on the client we leverage `RetryingBlockTransfer`/ `BlockTransferListener` to retry block push which relies on the `onFailure` callbacks.

Reply:
I have made the change to use a proper `RPCResponse` instead of the exception stack trace for the common block push failures.
Please review the updated change.